File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    q = deque()
    visited = [[0 for _ in range(len(maps[0]))] for _ in range(len(maps))]
    visited[0][0] = 1
    q.append((0, 0))

    while q:
        ci, cj = q.popleft()

        if ci == len(maps) - 1 and cj == len(maps[0]) - 1:
            return visited[ci][cj]

        for di, dj in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            ni = ci + di
            nj = cj + dj

            if (0 <= ni < len(maps) and 0 <= nj < len(maps[0]) and maps[ni][nj] == 1 and visited[ni][nj] == 0):
                q.append((ni, nj))
                visited[ni][nj] = visited[ci][cj] + 1

    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    class Pos{
        int i;
        int j;

        public Pos(int i, int j){
            this.i = i;
            this.j = j;
        }
    }

    public int bfs(Pos start, int[][] maps){
        Queue<Pos> q = new LinkedList<>();
        int n = maps.length;
        int m = maps[0].length;
        int[][] visited = new int[n][m];
        int[] dy = new int[]{1, -1, 0, 0};
        int[] dx = new int[]{0, 0, 1, -1};

        q.offer(start);
        visited[start.i][start.j] = 1;

        while(!q.isEmpty()){
            Pos cur = q.poll();

            if(cur.i == n-1 && cur.j == m-1){
                return visited[cur.i][cur.j];
            }

            for(int d = 0; d < 4; d++){
                int ni = cur.i + dy[d];
                int nj = cur.j + dx[d];

                if(0 <= ni && ni < n &&
                        0 <= nj && nj < m &&
                        maps[ni][nj] == 1 && visited[ni][nj] == 0){
                    q.offer(new Pos(ni, nj));
                    visited[ni][nj] = visited[cur.i][cur.j] + 1;
                }
            }
        }

        return -1;
    }

    public int solution(int[][] maps) {
        // 시작 위치는 (1, 1)이지만 2차원 배열에 적용하려면 (0, 0)이 되기 때문
        int answer = bfs(new Pos(0, 0), maps);
        
        return answer;
    }
}